src/ir_generator.rs

```diff
@@ -0,0 +1,193 @@
+use crate::compiler::Compiler;
+use crate::errors::{Severity, SourceError};
+use crate::parser::{AstNode, NodeId};
+use nu_protocol::ast::{Math, Operator};
+use nu_protocol::ir::{Instruction, IrBlock, Literal};
+use nu_protocol::{RegId, Span};
+
+/// Generates IR (Intermediate Representation) from nu AST.
+pub struct IrGenerator<'a> {
+    // Immutable reference to a compiler after the typechecker pass
+    compiler: &'a Compiler,
+    errors: Vec<SourceError>,
+
+    instructions: Vec<Instruction>,
+    register_count: u32,
+    file_count: u32,
+}
+
+impl<'a> IrGenerator<'a> {
+    pub fn new(compiler: &'a Compiler) -> Self {
+        Self {
+            compiler,
+            errors: Default::default(),
+            instructions: Default::default(),
+            register_count: 0,
+            file_count: 0,
+        }
+    }
+
+    /// Generates the IR from the given state of the compiler.
+    /// After this is called, use `block` and `errors` to get the result.
+    pub fn generate(&mut self) {
+        if self.compiler.ast_nodes.is_empty() {
+            return;
+        }
+        let Some(reg) = self.generate_node(NodeId(self.compiler.ast_nodes.len() - 1)) else {
+            return;
+        };
+        self.instructions.push(Instruction::Return { src: reg });
+    }
+
+    /// Returns generated IR block.
+    ///
+    /// Call `generate` before using this method and ensure there are no errors.
+    pub fn block(self) -> IrBlock {
+        // TODO: properly generate the spans
+        // TODO: figure out what to do with AST, as this parser has different
+        // representation of AST than the old parser.
+        let mut spans = vec![];
+        let mut ast = vec![];
+        for _ in 0..(self.instructions.len()) {
+            spans.push(Span { start: 0, end: 0 });
```

Comment:
Is there a reason to separate `generate()` and `block()` into two functions? I believe you could create the IrBlock at the beginning and then while generating the nodes, it could push the instructions to it with a proper span on the fly.

Better than using a Span is to modify IrBlock to use NodeId, but this can be done later and in stages (e.g., first use both Span and NodeId until we migrate fully, then deletee Span), not in the scope of this PR.

The AST field can be None since we don't need the old Nushell AST for anything. Once we use NodeId instead of Span, as outlined above, then the NodeIds form the sub-AST corresponding to the IrBlock, I think that should suffice.

Reply:
I believe `generate()` and `block()` are separated because of errors and to follow other classes (like typechecker). I could aldo just change `generate()` to return `Result<IrBlock, Vec<SourceError>>`. WDYT?

Applied the suggestion to create `IrBlock` in constructor, and then modify it as part of the execution.